cpp/src/interop/to_arrow_host.cu

```diff
@@ -0,0 +1,429 @@
+                     // simply matches the sign bit to satisfy the two's
+                     // complement integer representation of negative numbers.
+                     out[out_idx] = in[in_idx];
+#pragma unroll BIT_WIDTH_RATIO - 1
+                     for (auto i = 1; i < BIT_WIDTH_RATIO; ++i) {
+                       out[out_idx + i] = in[in_idx] < 0 ? -1 : 0;
+                     }
+                   });
+
+  return buf;
+}
+
+template std::unique_ptr<rmm::device_buffer> decimals_to_arrow<int32_t>(
+  cudf::column_view input, rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+
+template std::unique_ptr<rmm::device_buffer> decimals_to_arrow<int64_t>(
+  cudf::column_view input, rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+
+namespace {
+
+struct dispatch_to_arrow_host {
+  cudf::column_view column;
+  rmm::cuda_stream_view stream;
+  rmm::device_async_resource_ref mr;
+
+  int populate_validity_bitmap(ArrowBitmap* bitmap) const
+  {
+    if (!column.has_nulls()) { return NANOARROW_OK; }
+
+    NANOARROW_RETURN_NOT_OK(ArrowBitmapResize(bitmap, static_cast<int64_t>(column.size()), 0));
+    CUDF_CUDA_TRY(cudaMemcpyAsync(bitmap->buffer.data,
+                                  (column.offset() > 0)
+                                    ? cudf::detail::copy_bitmask(column, stream, mr).data()
+                                    : column.null_mask(),
+                                  bitmap->buffer.size_bytes,
+                                  cudaMemcpyDefault,
+                                  stream.value()));
+    return NANOARROW_OK;
+  }
+
+  template <typename T>
+  int populate_data_buffer(device_span<T const> input, ArrowBuffer* buffer) const
+  {
+    NANOARROW_RETURN_NOT_OK(ArrowBufferResize(buffer, input.size_bytes(), 1));
+    CUDF_CUDA_TRY(cudaMemcpyAsync(
+      buffer->data, input.data(), input.size_bytes(), cudaMemcpyDefault, stream.value()));
+    return NANOARROW_OK;
+  }
+
+  template <typename T,
+            CUDF_ENABLE_IF(!is_rep_layout_compatible<T>() &&
+                           (!std::is_same_v<T, numeric::decimal128> &&
+                            !std::is_same_v<T, numeric::decimal32> &&
+                            !std::is_same_v<T, numeric::decimal64>))>
+  int operator()(ArrowArray*) const
+  {
+    CUDF_FAIL("Unsupported type for to_arrow_host", cudf::data_type_error);
+  }
+
+  template <typename T,
+            CUDF_ENABLE_IF(is_rep_layout_compatible<T>() || std::is_same_v<T, numeric::decimal128>)>
+  int operator()(ArrowArray* out) const
+  {
+    nanoarrow::UniqueArray tmp;
+
+    auto const storage_type = id_to_arrow_storage_type(column.type().id());
+    NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), storage_type, column));
+
+    NANOARROW_RETURN_NOT_OK(populate_validity_bitmap(ArrowArrayValidityBitmap(tmp.get())));
+    using DataType = std::conditional_t<std::is_same_v<T, numeric::decimal128>, __int128_t, T>;
```

Comment:
Seems redundant to re-check with `std::conditional_t`

```suggestion
    using DataType = __int128_t;
```

Reply:
It's not being re-checked. This function is used for all the fixed-width types (except bool, decimal32 and decimal64). But `numeric::decimal128` is not `__int128_t`. So we need the conditional to check if this is the decimal128 implementation (in which case we use `__int128_t`) or any of the other versions, in which case we can just use `T`.

Reply:
I see. Ok makes sense.